### PR TITLE
Adding wrapping controller

### PR DESCRIPTION
**Pull request by AymericNoel**
I tried to code a first approach of the controller for the wrapping page. This will surely change because it's a first approach, while waiting to clarify the situation. I'd like to talk about it together.

### CODE REVIEW

File: src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+            var result = await _controller.GetTrakxAddressFromSymbol(symbol);
+            ((OkObjectResult) result.Result).Value.Should().Be(accountAddress);
+        }
+
+        [Fact]
+        public async Task GetTrakxAddressFromSymbol_should_send_badrequest_if_error_occurs_when_connecting_to_coinbase()
+        {
+            var symbol = "btc";
+            var accountAddress = "0x734Ac651Dd95a339c633cdEd410228515F97fAfF";
```
[the remark this answers is not in the document]
Comment (AymericNoel) [reply src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs:58]:
```suggestion
            
```

File: src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            _coinbaseClient.CustodiedCoins.Returns(new List<string> { "btc", "eth", "ltc" });
+            _wrappingInteraction.RetrieveAddressFromSymbol(symbol).Returns((string)null);
+
+            var result = await _controller.GetTrakxAddressFromSymbol(symbol);
+            ((BadRequestObjectResult)result.Result).Value.Should().Be("Sorry we can't find the address, please try again.");
+        }
+
+        [Fact]
+        public async Task WrapTokens_should_send_not_wrap_tokens_if_user_try_duplicate_one_transaction_or_didnt_make_any_transaction()
+        {
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs:68]: No big deal: I am not sure the name of the method reflects what this does, or I don't get the duplicate part... 🤷‍♂️

File: src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+using Trakx.Common.Sources.Coinbase;
+using Trakx.IndiceManager.Server.Controllers;
+using Trakx.IndiceManager.Server.Managers;
+using Trakx.IndiceManager.Server.Models;
+using Trakx.Tests.Data;
+using Xunit;
+
+namespace Trakx.IndiceManager.Server.Tests.Unit.Controllers
+{
+    public class WrappingControllerTest
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs:17]: Super nice and easy to read 👏👏

File: src/Trakx.IndiceManager.Server/Managers/IWrappingInteraction.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+using Trakx.Common.Interfaces.Transaction;
+using Trakx.IndiceManager.Server.Models;
+
+namespace Trakx.IndiceManager.Server.Managers
+{
+    /// <summary>
+    /// Use this component to retrieve information about Trakx' balance, Trakx addresses, Trakx transactions
+    /// from Coinbase custody and to send wrapped/unwrapped tokens.
+    /// </summary>
+    public interface IWrappingInteraction
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server/Managers/IWrappingInteraction.cs:15]: I am not a big fan of that name, but I agree it is not easy to find a good one 😅

File: src/Trakx.IndiceManager.Server/Models/WrappingTransactionModel.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            ReceiverAddress = transaction.ReceiverAddress;
+            Date = transaction.TimeStamp;
+            FromCurrency = transaction.FromCurrency;
+            ToCurrency = transaction.ToCurrency;
+            EthereumBlockId = transaction.EthereumBlockId;
+            NativeBlockId = transaction.NativeChainBlockId;
+            EthereumTransactionHash=transaction.NativeChainTransactionHash;
+            NativeChainTransactionHash= transaction.EthereumTransactionHash;
+        }
+        public string TrakxAddress { get; set; }
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server/Models/WrappingTransactionModel.cs:25]: I can't see this one being used 🤔

File: src/Trakx.IndiceManager.Server/Trakx.IndiceManager.Server.csproj [hunk 6]
```diff
@@ -24,6 +24,7 @@
     <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="3.1.3" />
     <PackageReference Include="Microsoft.OpenApi" Version="1.1.4" />
     <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.9.10" />
+    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.1.2" />
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server/Trakx.IndiceManager.Server.csproj:27]: Sorry I can't remember if we want that or not ?

File: src/Trakx.IndiceManager.Server/Models/WrappingTransactionModel.cs (hunk 5 above)
Comment (AymericNoel) [reply src/Trakx.IndiceManager.Server/Models/WrappingTransactionModel.cs:25]: In the wrapping tokens UI , we have the address where the user should send his native coins that is readonly. So in my head, the client request to the server this address which changes with the type of natives tokens in coinbase custody.

File: src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            
+            _coinbaseClient.CustodiedCoins.Returns(new List<string> { "btc", "eth", "ltc" });
+            _wrappingInteraction.RetrieveAddressFromSymbol(symbol).Returns((string)null);
+
+            var result = await _controller.GetTrakxAddressFromSymbol(symbol);
+            ((BadRequestObjectResult)result.Result).Value.Should().Be("Sorry we can't find the address, please try again.");
+        }
+
+        [Fact]
+        public async Task WrapTokens_should_send_not_wrap_tokens_if_user_try_duplicate_one_transaction_or_didnt_make_any_transaction()
```
Comment (AymericNoel) [inline src/Trakx.IndiceManager.Server.Tests/Unit/Controllers/WrappingControllerTest.cs:67]:
```suggestion
        public async Task WrapTokens_should_send_not_wrap_tokens_if_user_didnt_make_any_transaction()
```

File: src/Trakx.IndiceManager.Server/Trakx.IndiceManager.Server.csproj (hunk 6 above)
Comment (AymericNoel) [reply src/Trakx.IndiceManager.Server/Trakx.IndiceManager.Server.csproj:27]: I'll try to delete it to see what it does

File: src/Trakx.IndiceManager.Server/Managers/IWrappingInteraction.cs (hunk 4 above)
Comment (AymericNoel) [reply src/Trakx.IndiceManager.Server/Managers/IWrappingInteraction.cs:15]: IWrappingHelper ?
IWraapingTool?
IWrappingManager ?

Comment (monsieurleberre) [reply src/Trakx.IndiceManager.Server/Managers/IWrappingInteraction.cs:15]: 🤔 hard to say, maybe IWrappingService which we can divide into Wrapping Retriever and Wrapping Creator if it gets too big... WrappingService is not good either 😂 it is similarly too generic to be helpful but .... 🤷‍♂️ up to you honeslty

File: src/Trakx.IndiceManager.Server/Models/WrappingTransactionModel.cs (hunk 5 above)
Comment (monsieurleberre) [reply src/Trakx.IndiceManager.Server/Models/WrappingTransactionModel.cs:25]: OK I get it now, I just saw that no part of the code was referencing it, that is why I wondered ;)

**Pull request merged**